### PR TITLE
normalize juvare for il

### LINKED ISSUE #205: normalize juvare from IL
Blocked on #194.

[learn our pipeline: normalize](https://github.com/CAVaccineInventory/vaccine-feed-ingest/wiki/Runner-pipeline-stages#normalize)

Normalize existing [`.ndjson`](http://ndjson.org/) records into [the Vaccinate The States data schema](https://github.com/CAVaccineInventory/vaccine-feed-ingest/blob/main/vaccine_feed_ingest/schema/schema.py).

Read all files in the directory passed as the second argument (`sys.argv[2]`), transform each record into the normalized schema (filling in as much detail as possible), then output them to new files in the directory passed as the first argument (`sys.argv[1]`).

Check the wiki to learn more about the purpose of the normalize stage and how to get set up for development!

### Tips

1. Fetch and parse data for this site before you start developing:
    ```sh
    poetry run vaccine-feed-ingest fetch <state>/<site>
    poetry run vaccine-feed-ingest parse <state>/<site>
    ```

1. While working on your code, run it at any point:
    ```sh
    poetry run vaccine-feed-ingest normalize <state>/<site>
    ```

### Example
[Parsed files for `ak/arcgis`](https://github.com/CAVaccineInventory/vaccine-feed-ingest-results/tree/main/ak/arcgis/parsed) are converted to [normalized files](https://github.com/CAVaccineInventory/vaccine-feed-ingest-results/tree/main/ak/arcgis/normalized)

### PR DESCRIPTION
**Pull request by yotann**
# Normalize juvare for IL

| Key Details |
|-|
| Resolves #205 |
State: il |
Site: juvare |

## Notes
Several things about this normalizer need discussion:
- It's possible to generate direct links to the sign-up form for each site, like this: <https://events.juvare.com/IL-IDPH/0406ec6d-e7c9-40f0-aa86-98b0d8a5acdf/>. However, these links bypass the consent form at <https://covidvaccination.dph.illinois.gov/>, so I thought it would be safer to return that site instead.
- The source only seems to list sites that have available signup times. I'm therefore returning `"supply_level":"in_stock"`, on the assumption that any site with available times will have available stock.
- The source frequently includes duplicates of the same site with different dates or different vaccine types (1st dose vs. 2nd dose). I'm returning the duplicates as-is, but let me know if I should try eliminating them. This also means that `"opening_dates"` may only be a subset of the actual opening dates of the site.

## Data sample
```json
{"id": "il_juvare:285fe792-fcd0-4ca1-9471-dd92752c84b6", "name": "IDHS - Mabley Developmental Center Vaccination POD", "address": {"street1": "1120 Washington Avenue", "street2": null, "city": "Dixon", "state": "IL", "zip": "61201"}, "location": {"latitude": 41.861388, "longitude": -89.479465}, "contact": [{"contact_type": "booking", "phone": null, "website": "https://covidvaccination.dph.illinois.gov/", "email": null, "other": null}], "languages": null, "opening_dates": [{"opens": "2021-03-30", "closes": "2021-04-30"}], "opening_hours": null, "availability": {"drop_in": null, "appointments": true}, "inventory": null, "access": null, "parent_organization": null, "links": null, "notes": null, "active": true, "source": {"source": "il_juvare", "id": "285fe792-fcd0-4ca1-9471-dd92752c84b6", "fetched_from_uri": "https://covidvaccination.dph.illinois.gov/", "fetched_at": "2021-04-27T02:54:55.455657", "published_at": null, "data": {"organizer": "IL-IDPH", "slug": "285fe792-fcd0-4ca1-9471-dd92752c84b6", "name": "03/30 to 04/30 - IDHS - Mabley Developmental Center Vaccination POD", "location": "1120 Washington Avenue Dixon, IL 61201", "dateFrom": "3/30/2021, 9:00 AM", "dateTo": "4/30/2021, 3:30 PM", "lat": 41.861388, "lon": -89.479465, "search": "03/30 to 04/30 - idhs - mabley developmental center vaccination pod 1120 washington avenue dixon, il 61201 3/30/2021, 9:00 am 4/30/2021, 3:30 pm"}}}
{"id": "il_juvare:415331f6-0a4e-423a-b797-3c3aa99bb079", "name": "IDHS - Choate Developmental Center Vaccination", "address": {"street1": "1000 N. Main", "street2": null, "city": "Anna", "state": "IL", "zip": "62906"}, "location": {"latitude": 37.459465, "longitude": -89.197838}, "contact": [{"contact_type": "booking", "phone": null, "website": "https://covidvaccination.dph.illinois.gov/", "email": null, "other": null}], "languages": null, "opening_dates": [{"opens": "2021-03-31", "closes": "2021-05-19"}], "opening_hours": null, "availability": {"drop_in": null, "appointments": true}, "inventory": null, "access": null, "parent_organization": null, "links": null, "notes": null, "active": true, "source": {"source": "il_juvare", "id": "415331f6-0a4e-423a-b797-3c3aa99bb079", "fetched_from_uri": "https://covidvaccination.dph.illinois.gov/", "fetched_at": "2021-04-27T02:54:55.455657", "published_at": null, "data": {"organizer": "IL-IDPH", "slug": "415331f6-0a4e-423a-b797-3c3aa99bb079", "name": "03/31 to 05/19 - IDHS - Choate Developmental Center Vaccination", "location": "1000 N. Main, Anna, IL 62906", "dateFrom": "3/31/2021, 10:00 AM", "dateTo": "5/19/2021, 7:00 PM", "lat": 37.459465, "lon": -89.197838, "search": "03/31 to 05/19 - idhs - choate developmental center vaccination 1000 n. main, anna, il 62906 3/31/2021, 10:00 am 5/19/2021, 7:00 pm"}}}
{"id": "il_juvare:4f10bfeb-6d6a-4a9b-9059-1e06b405743f", "name": "State Mass Vaccination Site", "address": {"street1": "3128 Voyager Ln", "street2": null, "city": "Joliet", "state": "IL", "zip": "60435"}, "location": {"latitude": 41.575844, "longitude": -88.161001}, "contact": [{"contact_type": "booking", "phone": null, "website": "https://covidvaccination.dph.illinois.gov/", "email": null, "other": null}], "languages": null, "opening_dates": [{"opens": "2021-04-01", "closes": "2021-05-30"}], "opening_hours": null, "availability": {"drop_in": null, "appointments": true}, "inventory": null, "access": null, "parent_organization": null, "links": null, "notes": null, "active": true, "source": {"source": "il_juvare", "id": "4f10bfeb-6d6a-4a9b-9059-1e06b405743f", "fetched_from_uri": "https://covidvaccination.dph.illinois.gov/", "fetched_at": "2021-04-27T02:54:55.455657", "published_at": null, "data": {"organizer": "IL-IDPH", "slug": "4f10bfeb-6d6a-4a9b-9059-1e06b405743f", "name": "04/01/2021 \u2013 Will County \u2013 State Mass Vaccination Site", "location": "3128 Voyager Ln Joliet IL 60435", "dateFrom": "4/1/2021, 10:00 AM", "dateTo": "5/30/2021, 6:00 PM", "lat": 41.575844, "lon": -88.161001, "search": "04/01/2021 \u2013 will county \u2013 state mass vaccination site 3128 voyager ln joliet il 60435 4/1/2021, 10:00 am 5/30/2021, 6:00 pm"}}}
```

## Before Opening a PR
- [x] I tested this using the CLI (e.g., `poetry run vaccine-feed-ingest <state>/<site>`)
- [x] I ran auto-formatting: `poetry run tox -e lint-fix`

### CODE REVIEW

**Comment by eliblock** [comment]
Hi @yotann - thank you for submitting this! I probably won't knock out a full review this evening (someone else might, or I will tomorrow!), but I wanted to reply to your questions:

> It's possible to generate direct links to the sign-up form for each site, like this: https://events.juvare.com/IL-IDPH/0406ec6d-e7c9-40f0-aa86-98b0d8a5acdf/. However, these links bypass the consent form at https://covidvaccination.dph.illinois.gov/, so I thought it would be safer to return that site instead.

I think you're making the right call here. @obra - could you confirm?

> The source only seems to list sites that have available signup times. I'm therefore returning "supply_level":"in_stock", on the assumption that any site with available times will have available stock.

That seems perfect! This system is _not_ designed to track real-time availability of doses (which changes far faster than we can run the full pipeline), rather the `in_stock` (and other) are intended to point towards general trends. Your decision is inline with that rule of thumb.

> The source frequently includes duplicates of the same site with different dates or different vaccine types (1st dose vs. 2nd dose). I'm returning the duplicates as-is, but let me know if I should try eliminating them. This also means that "opening_dates" may only be a subset of the actual opening dates of the site.

While this is certainly frustrating, it is not unique to this dataset, and we are working on ways to handle duplicates-within-a-dataset downstream.

**Comment by yotann** [comment]
On direct links: I've noticed that several providers' websites link directly to the sign-up form and bypass the consent form. Examples: [Boone County](https://www.boonecountyil.org/government/departments/health_department/index.php), [BPM County](https://www.bchealthdepartment.org/). So maybe using the direct link is okay after all.

**Comment by obra** [comment]
If it's a *government provided* resource, I am 100% ok with us defaulting to the link they give. I do not object to us cleaning up the links to point to a consent form that links directly back to the resource provided.

If it's not a government site, slightly more caution is better.

**Comment by yotann** [comment]
The direct sign-up forms are on events.juvare.com, which belongs to a company contracted by the Illinois government to handle vaccine signups. The indirect consent form is on covidvaccination.dph.illinois.gov, but I think the underlying data is still managed by Juvare.

Actually, if I Google "Illinois vaccination event", one of the results is <https://events.juvare.com/IL-IDPH/> which also lets people bypass the consent form. They're really making it very easy to find the direct sign-up links without going through the consent form first. So I'll switch to returning the direct link.

File: vaccine_feed_ingest/runners/il/juvare/normalize.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+            schema.Vaccine(vaccine="johnson_johnson_janssen", supply_level="in_stock")
+        )
+
+    if len(inventory) == 0:
+        return None
+
+    return inventory
+
+
+def _filter_name(building: str, site: dict) -> str:
```
Comment (bryanculbertson) [inline vaccine_feed_ingest/runners/il/juvare/normalize.py:128]: minor: Can you provide some examples of original address -> name so I know the context of this?

**Review by bryanculbertson — APPROVED** [review]
This is an impressive normalizer! Bravo!

**Review by tmb — CHANGES_REQUESTED** [review]
Just one thing — is it possible to include the `opening_hours` datapoint for the clinics that are only open for one day? I know it's impractical for the ones that are like a month long, but for ones like these:

```
{"id": "il_juvare:qqvjg-8", "name": "Victory Apostolic Church", "address": {"street1": "20801 Matteson Ave.", "street2": null, "city": "Matteson", "state": "IL", "zip": "60443"}, "location": {"latitude": 41.51731, "longitude": -87.73767}, "contact": [{"contact_type": "booking", "phone": null, "website": "https://events.juvare.com/IL-IDPH/qqvjg-8/", "email": null, "other": null}], "languages": null, "opening_dates": [{"opens": "2021-05-01", "closes": "2021-05-01"}], "opening_hours": null, "availability": {"drop_in": null, "appointments": true}, "inventory": [{"vaccine": "moderna", "supply_level": "in_stock"}], "access": null, "parent_organization": null, "links": null, "notes": null, "active": true, "source": {"source": "il_juvare", "id": "qqvjg-8", "fetched_from_uri": "https://covidvaccination.dph.illinois.gov/", "fetched_at": "2021-04-29T20:53:08.732027", "published_at": null, "data": {"organizer": "IL-IDPH", "slug": "qqvjg-8", "name": "05/01/2021 - Emerge MedStaffing - Victory Apostolic Church Vaccination Clinic - Moderna", "location": "20801 Matteson Ave. Matteson, IL 60443\r\nVictory Apostolic Church", "dateFrom": "5/1/2021, 9:00 AM", "dateTo": "5/1/2021, 4:00 PM", "lat": 41.51731, "lon": -87.73767, "search": "05/01/2021 - emerge medstaffing - victory apostolic church vaccination clinic - moderna 20801 matteson ave. matteson, il 60443\r\nvictory apostolic church 5/1/2021, 9:00 am 5/1/2021, 4:00 pm"}}}
```

maybe it would work?

Let me know what you think — if it's uncommon enough in the data then I'm happy to go ahead w merge now :)

**Comment by eliblock** [comment]
Hi! We've added significant validation to the schema, and want to make sure this normalizer takes advantage of it.

Could you pull in the latest `main` and run your code on top of it? Full details on what changed and why are in #278

File: vaccine_feed_ingest/runners/il/juvare/normalize.py (hunk 1 above)
Comment (yotann) [reply vaccine_feed_ingest/runners/il/juvare/normalize.py:175]: Done.

**Comment by yotann** [comment]
I've made it work with the new schema, and added comments and examples. @tmb I did add opening hours as you suggested. It would be possible to get the opening hours for every location, but that would require a fair amount of extra work, so I opened a new issue (#297).

**Comment by eliblock** [comment]
Thank you for the updates to the improved schema, and for opening the followup issue!

**Pull request merged**